File: hash/good_subseqs_sum/good_subseqs_sum.py

def sum_good_subsequences(nums: list[int]) -> int:  # LeetCode Q.3351.
    good_subseqs_sum = 0
    subsequences_table: dict[int, list[int]] = dict()
    for num in nums:
        if num not in subsequences_table.keys():
            subsequences_table.update(
                {num: [0, 0]}  # Format: [count, subseqs sum].
            )
        subsequences_table[num][0] += 1
        subsequences_table[num][1] += num
        good_subseqs_sum += num

        # Current num extends subseqs at such two ends.
        for subseq_end in (num - 1, num + 1):
            if subseq_end in subsequences_table.keys():
                subsequences_table[num][0] += subsequences_table[subseq_end][0]

                new_sum = subsequences_table[subseq_end][1]
                new_sum += subsequences_table[subseq_end][0] * num

                subsequences_table[num][1] += new_sum
                good_subseqs_sum += new_sum

    return good_subseqs_sum % (10 ** 9 + 7)  # Required to control size.

File: hash/good_subseqs_sum/good_subseqs_sum.cpp
#include <vector>
#include <unordered_map>
using namespace std;

int sumGoodSubsequences(vector<int> &nums) // LeetCode Q.3351.
{
    int modulo = pow(10, 9) + 7;

    int64_t goodSubseqsSum = 0; // 64-bit int: avoid overflow.

    // Keys: subseqs ending at num; values: such subseqs' count and elements sum.
    unordered_map<int, vector<int64_t>> subseqsTable = {};

    for (auto num : nums)
    {
        if (subseqsTable.find(num) == subseqsTable.end())
            subseqsTable[num] = {0, 0}; // Format: {count, elements sum}.

        subseqsTable[num][0] += 1;
        subseqsTable[num][1] += num;
        goodSubseqsSum += num;

        // Current num extends subseqs at such two ends.
        for (auto subseqEnd : {num - 1, num + 1})
        {
            if (subseqsTable.find(subseqEnd) != subseqsTable.end())
            {
                subseqsTable[num][0] += subseqsTable[subseqEnd][0];

                int64_t newSum = subseqsTable[subseqEnd][1];
                newSum += num * subseqsTable[subseqEnd][0];

                subseqsTable[num][1] += newSum;
                goodSubseqsSum += newSum;
            }
        }

        subseqsTable[num][0] %= modulo; // Avoid overflow.
        subseqsTable[num][1] %= modulo;
        goodSubseqsSum %= modulo;
    }

    return goodSubseqsSum;
}